plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveDynamicPartitionPruningTest.java

```diff
@@ -37,6 +47,62 @@ protected QueryRunner createQueryRunner()
                 .build();
     }
 
+    @Test(timeOut = 30_000, dataProvider = "joinDistributionTypes")
+    public void testJoinDynamicFilteringMultiJoin(FeaturesConfig.JoinDistributionType joinDistributionType)
+    {
+        assertUpdate("DROP TABLE IF EXISTS t0");
+        assertUpdate("DROP TABLE IF EXISTS t1");
+        assertUpdate("DROP TABLE IF EXISTS t2");
+        assertUpdate("CREATE TABLE t0 (v0 real, k0 integer) WITH (partitioned_by=array['k0'])");
+        assertUpdate("CREATE TABLE t1 (v1 real, i1 integer)");
+        assertUpdate("CREATE TABLE t2 (v2 real, i2 integer, k2 integer) WITH (partitioned_by=array['k2'])");
+        assertUpdate("INSERT INTO t0 VALUES (1.0, 1), (1.0, 2)", 2);
+        assertUpdate("INSERT INTO t1 VALUES (2.0, 10), (2.0, 20)", 2);
+        assertUpdate("INSERT INTO t2 VALUES (3.0, 1, 1), (3.0, 2, 2)", 2);
+
+        // queries should not deadlock
+
+        // t0 table scan depends on DFs from t1 and t2
+        assertDynamicFilters(
+                noJoinReordering(joinDistributionType),
+                "SELECT v0, v1, v2 FROM (t0 JOIN t2 ON k0 = i2) JOIN t1 ON k0 = i1");
+
+        // DF evaluation order is: t1 => t2 => t0
+        assertDynamicFilters(
+                noJoinReordering(joinDistributionType),
+                "SELECT v0, v1, v2 FROM (t0 JOIN t2 ON k0 = i2) JOIN t1 ON k2 = i1");
+
+        // t2 table scan depends on t1 DFs, but t0 <-> t2 join is blocked on t2 data
+        assertDynamicFilters(
+                noJoinReordering(joinDistributionType),
+                "SELECT v0, v1, v2 FROM (t0 JOIN t2 ON (k0 * -1) + 2 = i2) JOIN t1 ON k2 = i1");
```

Comment:
It would be good to have a comment here that the expression `(k0 * -1) + 2 = i2` is being used instead of `k0 = i2` to avoid `t0` table scan having dependency on a DF